### PR TITLE
rfc: "No action" / default property value standardization

### PR DESCRIPTION
**Pull request by andrefcdias**
## New Behavior

This RFC aims to standardize what values we use in our components for cases where a prop value results in no action, i.e. defaults that apply no styles.

Currently, the approach we follow for all components is to use a string value like 'off', 'none' or 'default' for default values of a prop.

[PREVIEW 🔍](https://github.com/andrefcdias/fluentui/blob/rfc-standardize-no-action-prop-values/rfcs/react-components/no-action-prop-value-standardization.md)

## Related Issue(s)

#23254

### CODE REVIEW

File: rfcs/react-components/no-action-prop-value-standardization.md [hunk 1]
```diff
@@ -0,0 +1,80 @@
...
+Cases where our component should have a default value, we would apply the state that should be the default value.
+i.e. `Card` component's `appearance` prop, should have a default appearance, as we don't want a card without a pre-defined appearance as that's detrimental to the usage.
+
+Cases where nothing is done in our hooks as a default, we would leverage `undefined` and `null`.
+i.e. `Card` component's `focusMode` prop should be nullable instead of using an `off` string value, as that reflects the fact that nothing is applied to the component.
+
+### Pros and Cons
+
+#### Pros
+
```
[the remark this answers is not in the document]
Comment (Hotell) [reply rfcs/react-components/no-action-prop-value-standardization.md:48]: maybe a topic for a follow-up RFC, but we could also adopt more straightforward default patterns logic with less code (when using `undefined` to express "default" state. 
image

File: rfcs/react-components/no-action-prop-value-standardization.md [hunk 2]
```diff
@@ -0,0 +1,80 @@
...
+## Detailed Design or Proposal
+
+This RFC proposes that we leverage the standard JavaScript default, `undefined`, for attributes instead of a string.
+For this, we should stop using `Required` for all props in the `State` and decide case by case if something should / should not be optional.
+
+Cases where our component should have a default value, we would apply the state that should be the default value.
+i.e. `Card` component's `appearance` prop, should have a default appearance, as we don't want a card without a pre-defined appearance as that's detrimental to the usage.
+
+Cases where nothing is done in our hooks as a default, we would leverage `undefined` and `null`.
+i.e. `Card` component's `focusMode` prop should be nullable instead of using an `off` string value, as that reflects the fact that nothing is applied to the component.
```
Comment (Hotell) [inline rfcs/react-components/no-action-prop-value-standardization.md:43]: can we provide a "demo" example ( actual diff of this change) ?

File: rfcs/react-components/no-action-prop-value-standardization.md [hunk 3]
```diff
@@ -0,0 +1,80 @@
...
+
+## Detailed Design or Proposal
+
+This RFC proposes that we leverage the standard JavaScript default, `undefined`, for attributes instead of a string.
+For this, we should stop using `Required` for all props in the `State` and decide case by case if something should / should not be optional.
+
+Cases where our component should have a default value, we would apply the state that should be the default value.
+i.e. `Card` component's `appearance` prop, should have a default appearance, as we don't want a card without a pre-defined appearance as that's detrimental to the usage.
+
+Cases where nothing is done in our hooks as a default, we would leverage `undefined` and `null`.
```
[the remark this answers is not in the document]
Comment (andrefcdias) [reply rfcs/react-components/no-action-prop-value-standardization.md:42]: I think that's appropriate; I'll change it but keep this conversation open for anyone who opposes the idea

File: rfcs/react-components/no-action-prop-value-standardization.md (hunk 2 above)
Comment (andrefcdias) [reply rfcs/react-components/no-action-prop-value-standardization.md:43]: Sure, the changes to the component hook and style hook?

File: rfcs/react-components/no-action-prop-value-standardization.md (hunk 1 above)
[the remark this answers is not in the document]
Comment (andrefcdias) [reply rfcs/react-components/no-action-prop-value-standardization.md:48]: Can you elaborate on what you mean by shape?

[the remark this answers is not in the document]
Comment (andrefcdias) [reply rfcs/react-components/no-action-prop-value-standardization.md:48]: > maybe a topic for a follow-up RFC, but we could also adopt more straightforward default patterns logic with less code (when using `undefined` to express "default" state. image

Just as an FYI, `Text` is currently not aligned with the rest of the components in terms of default value setting. Most components use defaults on the deconstruction of `props`.

File: rfcs/react-components/no-action-prop-value-standardization.md [hunk 4]
```diff
@@ -0,0 +1,80 @@
...
+
+#### Pros
+
+- Follows JavaScript standards
+- Intuitive to the user because of the above pro - no need to read documentation
+- Consistency across our product - enables ease of use
+
+#### Cons
+
+- New style map approach would require changes
```
[the remark this answers is not in the document]
Comment (andrefcdias) [reply rfcs/react-components/no-action-prop-value-standardization.md:55]: The example I provided is the status quo. I have no proposal for the solution of this con as of yet.
I'll add yours as a possible solution.

File: rfcs/react-components/no-action-prop-value-standardization.md [hunk 5]
```diff
@@ -0,0 +1,97 @@
...
+  wrap: true,
+});
+```
+
+## Detailed Design or Proposal
+
+This RFC proposes that we leverage the standard JavaScript default, `undefined`, for attributes instead of a string.
+For this, we should stop using `Required` for all props in the `State` and decide case by case if something should / should not be optional.
+
+Cases where our component should have a default value, we would apply the state that should be the default value.
```
Comment (ling1726) [inline rfcs/react-components/no-action-prop-value-standardization.md:39]: does this mean that for discriminate unions with two types we should actually convert it to a prop to a single boolean value ?

````tsx`
// before
<Button iconPosition="before" />
<Button iconPosition="after" />

//after
<Button iconAfter />
```

or do you mean that `undefined` in this case just implies that one of the options is used?

Comment (andrefcdias) [reply rfcs/react-components/no-action-prop-value-standardization.md:39]: Second option. What I mean here is that for props where the default value has side effects, a value, of any type, should be used as default instead of persisting an `undefined` to the state.
For example, with Card, we don't want a case where a user does not apply any `appearance`. As such, 'filled' is used as a default and passed into the state.
However, for example `focusMode`, when this value is `undefined`, nothing is done to the component. As such, using the underlying `undefined`, instead of defaulting to something like `"off"`, allows us to have a standard behavior across all components.

Comment (ling1726) [reply rfcs/react-components/no-action-prop-value-standardization.md:39]: AH ok that makes sense could you please make that more explicit (maybe with a code example) in the RFC ? it was not clear for me

Comment (andrefcdias) [reply rfcs/react-components/no-action-prop-value-standardization.md:39]: On it 👍🏻

**Review by ling1726 — APPROVED** [review]

File: rfcs/react-components/no-action-prop-value-standardization.md [hunk 6]
```diff
@@ -0,0 +1,190 @@
...
+  const className = mergeClasses(
+    classNames.root,
+    styles.root,
+    alignMap[props.align ?? '_default']
+  );
+  ```
+
+- Users who dynamically set a property value would need to do something like:
+  ```tsx
+  <Text align={isItFriday ? 'center' : undefined} />
```
Comment (ling1726) [inline rfcs/react-components/no-action-prop-value-standardization.md:177]: What is the 'before' version of this ? I'm not sure how this affects props

**Comment by bsunderhus** [comment]
Can you provide an example of a current component that needs this kind of modification right now @andrefcdias? I'm still not 100% sure I'm following your modifications request.

The way I see the value returned from `useComponent_unstable` state hook should return a well-defined single source of truth state, and having possible `undefined` values will disassociate logic from state hook into other hooks (in this case `useComponentStyle_unstable` hook)

File: rfcs/react-components/no-action-prop-value-standardization.md (hunk 6 above)
Comment (andrefcdias) [reply rfcs/react-components/no-action-prop-value-standardization.md:177]: The before would be something like:

```tsx
<Text align={isItFriday ? 'center' : 'none'} />
```

Dynamically setting the property will require an explicit passing of an `undefined` value.

File: rfcs/react-components/no-action-prop-value-standardization.md (hunk 1 above)
[the remark this answers is not in the document]
Comment (Hotell) [reply rfcs/react-components/no-action-prop-value-standardization.md:145]: > Can you elaborate on what you mean by shape

sorry for confusion, as TS is structural type system everything is basically a kind of a shape (interface, type alias as dictionary etc)

**Comment by andrefcdias** [comment]
> Can you provide an example of a current component that needs this kind of modification right now @andrefcdias? I'm still not 100% sure I'm following your modifications request.
> 
> The way I see the value returned from `useComponent_unstable` state hook should return a well-defined single source of truth state, and having possible `undefined` values will disassociate logic from state hook into other hooks (in this case `useComponentStyle_unstable` hook)
> 
> If the state value should include the value `undefined` then maybe using optional parameters is not ideal, perhaps we should be explicitly adding `undefined` as a type. That would mean that although `Required` is being used, `undefined` is still a valid value.
> 
> ```tsx
> interface OptionalType {
>   optional?: string
> }
> interface NonOptionalType {
>   optional: string | undefined
> }
> 
> const t: Required<OptionalType> = {optional: undefined} // error 🚨
> const t2: Required<NonOptionalType> = {optional: undefined} // fine ✅
> ```

This RFC originated from addressing this issue #22855 (PR: #23254). I noted that there wasn't a standard across our components for what the default value should be for when we do nothing with said value. While in `Card` I'm using `'off'` for `focusMode`, in `Text` using `'off'` would not work as it doesn't make sense for something like fonts (`font="off"`). Other components, for example Avatar's `active` prop, use the value `"unset"`, even though this isn't applied the CSS value of "unset" or anything.

It's a fact that there's no standard default value and that quite possibly there's no string value to rule them all, by fitting all use cases. As such, JS standards seem like the appropriate approach to unify behavior and improve DX by avoiding unnecessary documentation reading.

I think that your solution is good as it addresses the problem of standardizing what value represents a no action scenario. However, that comes at the cost of forcing users to initialize a state with all props with the value `undefined`.

```tsx
interface OptionalType {
  optional1?: string
  optional2?: string
  optional3?: string
  optional4?: string
  optional5?: string
}
interface NonOptionalType {
  optional1: string | undefined
  optional2: string | undefined
  optional3: string | undefined
  optional4: string | undefined
  optional5: string | undefined
}

const t: OptionalType = { // fine ✅
  optional3: "cool-value"
} 
const t2: NonOptionalType = { // error 🚨
  optional3: "cool-value"
}
const t3: WorkingNonOptionalType = { // fine ✅
  optional1: undefined,
  optional2: undefined,
  optional3: "cool-value"
  optional4: undefined,
  optional5: undefined
}
```

File: rfcs/react-components/no-action-prop-value-standardization.md (hunk 1 above)
[the remark this answers is not in the document]
Comment (andrefcdias) [reply rfcs/react-components/no-action-prop-value-standardization.md:145]: Ah, great point! I'll add that in.

File: rfcs/react-components/no-action-prop-value-standardization.md [hunk 7]
```diff
@@ -0,0 +1,191 @@
...
+state.root.className = mergeClasses(
+  cardClassNames.root,
+  styles.root,
+  appearanceLookup[state.appearance],
+  // {...}
+  state.root.className,
+);
+```
+
+Cases where nothing is done in our hooks as a default, we should leverage `undefined`.
```
Comment (Hotell) [inline rfcs/react-components/no-action-prop-value-standardization.md:77]:
```suggestion
Cases where nothing is done in our hooks as a default, we should leverage JavaScript standard for no value - `undefined`.
```

File: rfcs/react-components/no-action-prop-value-standardization.md [hunk 8]
```diff
@@ -0,0 +1,191 @@
...
+i.e. `Text` component's `font` prop should be nullable instead of using a `'default'` string value, as that reflects the fact that there are no side-effects from it.
```
Comment (Hotell) [inline rfcs/react-components/no-action-prop-value-standardization.md:78]: maybe we should improve terminology as `nullable` is not what we are proposing. 

```
`Text` component's `font` prop should be optional/undefined instead of using a `'default'` 
```

thoughts?

Comment (andrefcdias) [reply rfcs/react-components/no-action-prop-value-standardization.md:78]: C# must've been knocking on my door when I wrote this 🤔 Will make the suggested change.

File: rfcs/react-components/no-action-prop-value-standardization.md [hunk 9]
```diff
@@ -0,0 +1,191 @@
...
+// useText.ts
+const { font /* {...} */ } = props; // We no longer set the default here
+
+const state = {
+  font,
+  // {...}
+};
+
+// useTextStyles.ts
+const { font = '_default' } = state; // Applying a default to keep using a lookup
```
[the remark this answers is not in the document]
Comment (khmakoto) [reply rfcs/react-components/no-action-prop-value-standardization.md:99]: I see that you're proposing how to do that below, it probably needs to be changed here too though.

**Review by behowell — CHANGES_REQUESTED** [review]
I don't think we should make this change. 

The main motivation from the RFC seems to be to make the `useFooStyles_unstable` hooks easier to use on their own, by not requiring default values to be given. However, they're not meant to be used on their own. They should be given a state object that was initialized by the `useFoo_unstable` hook, which should be the single source of truth for the default values.

If there are other motivations, could you please add them to the RFC?

My main arguments to keep the status-quo are:
* For all public APIs (aka props), we are _already_ using `undefined` to mean "use the default value". And the default value is included in the documentation.
* We haven't settled on the hooks re-composition pattern yet. In some proposals, there is no publicly available `useFooStyles`-type hook (just a single component hook that returns state, including all styling applied). I don't think we should make changes just for the style hooks at this point, if they might never be made public anyways.
* The default values are all provided explicitly by the useFoo hooks. This gives a single place that defines the default behavior. 
  * With the proposal, all other parts of the code need to have their own interpretation of what undefined means. It also makes that code more complex by adding additional checks for `undefined`, or inventing a `'_default'` name to use locally.
* Using `Required` for state props helps catch coding mistakes. E.g. if the dev forgot to include the prop in the state object, or didn't provide a default value when they should have. Sure, we can catch that kind of thing through testing, but IMO it's always better to catch mistakes at compile time if possible.

**Comment by andrefcdias** [comment]
> I don't think we should make this change.
> 
> The main motivation from the RFC seems to be to make the `useFooStyles_unstable` hooks easier to use on their own, by not requiring default values to be given. However, they're not meant to be used on their own. They should be given a state object that was initialized by the `useFoo_unstable` hook, which should be the single source of truth for the default values.
> 
> If there are other motivations, could you please add them to the RFC?
> 
> My main arguments to keep the status-quo are:
> 
> * For all public APIs (aka props), we are _already_ using `undefined` to mean "use the default value". And the default value is included in the documentation.
> * We haven't settled on the hooks re-composition pattern yet. In some proposals, there is no publicly available `useFooStyles`-type hook (just a single component hook that returns state, including all styling applied). I don't think we should make changes just for the style hooks at this point, if they might never be made public anyways.
> * The default values are all provided explicitly by the useFoo hooks. This gives a single place that defines the default behavior.
>   
>   * With the proposal, all other parts of the code need to have their own interpretation of what undefined means. It also makes that code more complex by adding additional checks for `undefined`, or inventing a `'_default'` name to use locally.
> * Using `Required` for state props helps catch coding mistakes. E.g. if the dev forgot to include the prop in the state object, or didn't provide a default value when they should have. Sure, we can catch that kind of thing through testing, but IMO it's always better to catch mistakes at compile time if possible.

@behowell 

This RFC aims to achieve a standardization of the default values as, as stated in it, there's a plethora of words used and, if we can achieve a standard, the user doesn't need to check documentation.
Also, as stated by some users, these default values can be misleading, like "unset" in Avatar's `active` prop which does nothing to the component but is a very important keyword in the CSS world that indicates a reset.
This is why I'm proposing not using a default string value but instead reusing the naturally explicit value of `undefined` as it is a standard across JavaScript development.
From my point of view, if a non-undefined value is an option, it naturally has an action.
i.e in an `appearance` prop, `undefined` resolves to a string value that will apply a default style whilst in a prop like Card's `focusMode`, `undefined` resolves to an `"off"` value that **should** do something, but does not.
If not leverage `undefined` then coming up with a standard string value that fits all cases seems to me like the only other option.

Also, the current discussions have been around the consequences of the proposal, which is definitely not a blocker for the main topic, so feel free to propose other ideas on how to address the underlying naming problem of default values to avoid the cons you mentioned.
Please also let me know how I can clarify the RFC to reduce confusion over what it's trying to achieve as it's proving to be a common misconception.

**Comment by layershifter** [comment]
Can I try to summarise this RFC to two statements? (to ensure that my understanding is correct)

- When a component has a default state/behavior, it SHOULD HAVE a default value.

   _Example:_ `Card.appearance` has `'filled' | 'filled-alternative' | 'outline' | 'subtle'` and is `filled` by default
- When a component does not have a default state/behavior, it SHOULD NOT HAVE a default value.

  _Anti-example (current state)_: `Text.font` has `'base' | 'monospace' | 'numeric'` and defaults to `base`, where `base` **does not** apply any styles
  _Example (proposal)_: `Text.font` has `'monospace' | 'numeric'` and has no defaults (i.e. `undefined`)

---

@andrefcdias Is it accurate? If so I think that it's a good change that we need to make.

Where I agree with @behowell that we may not need changes to `use*Styles_unstable` hooks as they are unstable ¯\_(ツ)_/¯

> This RFC proposes that we leverage the standard JavaScript default, undefined, for attributes instead of a string. For this, we should stop using Required for all props in the State and decide case by case if something should / should not be optional.

Another approach to consider is something like [RequiredAllowingUndefined](https://www.typescriptlang.org/play?ssl=7&ssc=6&pln=7&pc=31#code/C4TwDgpgBACgTgezAZygXigbwFBSgMwQDtgB+ALigHJCEqoAfagIwEM56mBXIgEwnwBLIhF7YAvtmyhIUAMrBWwaBgBKEAI5dBcUQB54SZAD4pM6Oq07RAQQA2dhAHdhAcwCqfAcP0AVY+hYuHhQANoA0lDCUADWECAI+FCW2rq8ev4AupS+EZkSANxSAPTFUAACwMgAtBAAHpAAxsC1cIhwUABEPPxCIrydUahECMBQrA7OotiNxMhj80oQAIyUCkuBmATEwJQ93v1QkqUVVbUNEM2t7V2EJIOCw6NQvT5is0TzUIvKAExrimUm0kUg+Xx+EAAzJQUtZePZHC4iB4vH19IYUAEMFs7rsoPs0bwjtgTpUavUmi0IG0EB1OriHk8xq9+jM5gtARAACwwzSpWyTJEolnoxCY4HYIA). In this case we will still need to enumerate all properties on `state`, but `undefined` will be allowed:

```ts
const state3: RequiredAllowingUndefined<Props> = { font: undefined }
```

_However_ I don't see obvious benefits in this approach as it will only increase bundle size.

**Comment by andrefcdias** [comment]
> Can I try to summarise this RFC to two statements? (to ensure that my understanding is correct)
> 
> * When a component has a default state/behavior, it SHOULD HAVE a default value.
>   _Example:_ `Card.appearance` has `'filled' | 'filled-alternative' | 'outline' | 'subtle'` and is `filled` by default
> * When a component does not have a default state/behavior, it SHOULD NOT HAVE a default value.
>   _Anti-example (current state)_: `Text.font` has `'base' | 'monospace' | 'numeric'` and defaults to `base`, where `base` **does not** apply any styles
>   _Example (proposal)_: `Text.font` has `'monospace' | 'numeric'` and has no defaults (i.e. `undefined`)
> 
> @andrefcdias André Dias (he/him) FTE Is it accurate? If so I think that it's a good change that we need to make.
> 
> Where I agree with @behowell Ben Howell FTE that we may not need changes to `use*Styles_unstable` hooks as they are unstable ¯_(ツ)_/¯
> 
> > This RFC proposes that we leverage the standard JavaScript default, undefined, for attributes instead of a string. For this, we should stop using Required for all props in the State and decide case by case if something should / should not be optional.
> 
> Another approach to consider is something like [RequiredAllowingUndefined](https://www.typescriptlang.org/play?ssl=7&ssc=6&pln=7&pc=31#code/C4TwDgpgBACgTgezAZygXigbwFBSgMwQDtgB+ALigHJCEqoAfagIwEM56mBXIgEwnwBLIhF7YAvtmyhIUAMrBWwaBgBKEAI5dBcUQB54SZAD4pM6Oq07RAQQA2dhAHdhAcwCqfAcP0AVY+hYuHhQANoA0lDCUADWECAI+FCW2rq8ev4AupS+EZkSANxSAPTFUAACwMgAtBAAHpAAxsC1cIhwUABEPPxCIrydUahECMBQrA7OotiNxMhj80oQAIyUCkuBmATEwJQ93v1QkqUVVbUNEM2t7V2EJIOCw6NQvT5is0TzUIvKAExrimUm0kUg+Xx+EAAzJQUtZePZHC4iB4vH19IYUAEMFs7rsoPs0bwjtgTpUavUmi0IG0EB1OriHk8xq9+jM5gtARAACwwzSpWyTJEolnoxCY4HYIA). In this case we will still need to enumerate all properties on `state`, but `undefined` will be allowed:
> 
> ```ts
> const state3: RequiredAllowingUndefined<Props> = { font: undefined }
> ```
> 
> _However_ I don't see obvious benefits in this approach as it will only increase bundle size.

That is a great summary to it @layershifter, I'll be adding it to the RFC as it's much better explained and to the point. Thanks!

**Comment by behowell** [comment]
Ok, so as I understand now, the main thing to decide here is about the public API: whether we include a named value for a prop's default, or only have `undefined`. Does that sound right? The other parts about the implementation of the State type and useStyles hooks seem like secondary decisions.

I think this boils down to a stylistic choice. And with all stylistic choices, both options have valid reasons for and against, and can be a source of endless debate 🙂.

I still generally think it's better to leave things as they are (more reasoning below), but I understand if you disagree. I'm going to be on vacation for the next two weeks, so I won't be able to respond during that time. I'll remove my review from the PR to not block further progress. 

---

In general, I think when a prop has a union value, the API is clearer if you have names for all possible variants including the default one. That way it's more straightforward to document what happens for each variant: the documentation lists all of the possible values, and then says which one is the default. I would usually expect that `undefined` would map to one of the available values, rather than itself being a unique variant that's different from the other values... but that's where the stylistic debate comes in.

We could review the "off"/"none"/etc. type values and try to standardize them, but I don't think that full alignment across all props is necessary, since they don't all mean the same thing anyways. For example, something not being focusable `focusMode="off"` is different from something using the default font family `font="base"`. So there's no reason that they should have the same name.

I'm not entirely opposed to using `undefined` for some props. But I'm not sure that we should try to make a blanket policy about using `undefined` for the default value. Especially when the criteria for when to do it are a little hard to nail down... For example, regarding @layershifter's point here:
> * When a component does not have a default state/behavior, it SHOULD NOT HAVE a default value.
>   _Anti-example (current state)_: `Text.font` has `'base' | 'monospace' | 'numeric'` and defaults to `base`, where `base` **does not** apply any styles
>   _Example (proposal)_: `Text.font` has `'monospace' | 'numeric'` and has no defaults (i.e. `undefined`)

This feels like the criteria is being based on an implementation detail. For what it's worth, I'd say that `font="base"` _does_ in fact apply styles. The component author chose to put them on the `root` class, and let the other font values override that, rather than having a separate `base` class. That's very reasonable, but it could have been written like this instead:
```diff
  root: {
-    fontFamily: tokens.fontFamilyBase,
    // ...
  },
+  base: { fontFamily: tokens.fontFamilyBase },
  monospace: { fontFamily: tokens.fontFamilyMonospace },
  numeric: { fontFamily: tokens.fontFamilyNumeric },
```
```diff
  state.root.className = mergeClasses(
    styles.root,
+    state.font === 'base' && styles.base,
    state.font === 'monospace' && styles.monospace,
    state.font === 'numeric' && styles.numeric,
    // ...
  );
```
That's not to say that's a preferable way to do it, but just to illustrate that it is an implementation detail whether a given value applies styles or not. And implementation details shouldn't drive the API design.

**Comment by andrefcdias** [comment]
> Ok, so as I understand now, the main thing to decide here is about the public API: whether we include a named value for a prop's default, or only have `undefined`. Does that sound right? The other parts about the implementation of the State type and useStyles hooks seem like secondary decisions.
> 
> I think this boils down to a stylistic choice. And with all stylistic choices, both options have valid reasons for and against, and can be a source of endless debate 🙂.
> 
> I still generally think it's better to leave things as they are (more reasoning below), but I understand if you disagree. I'm going to be on vacation for the next two weeks, so I won't be able to respond during that time. I'll remove my review from the PR to not block further progress.
> 
> In general, I think when a prop has a union value, the API is clearer if you have names for all possible variants including the default one. That way it's more straightforward to document what happens for each variant: the documentation lists all of the possible values, and then says which one is the default. I would usually expect that `undefined` would map to one of the available values, rather than itself being a unique variant that's different from the other values... but that's where the stylistic debate comes in.
> 
> We could review the "off"/"none"/etc. type values and try to standardize them, but I don't think that full alignment across all props is necessary, since they don't all mean the same thing anyways. For example, something not being focusable `focusMode="off"` is different from something using the default font family `font="base"`. So there's no reason that they should have the same name.
> 
> I'm not entirely opposed to using `undefined` for some props. But I'm not sure that we should try to make a blanket policy about using `undefined` for the default value. Especially when the criteria for when to do it are a little hard to nail down... For example, regarding @layershifter Oleksandr Fediashov aka shift FTE's point here:
> 
> > * When a component does not have a default state/behavior, it SHOULD NOT HAVE a default value.
> >   _Anti-example (current state)_: `Text.font` has `'base' | 'monospace' | 'numeric'` and defaults to `base`, where `base` **does not** apply any styles
> >   _Example (proposal)_: `Text.font` has `'monospace' | 'numeric'` and has no defaults (i.e. `undefined`)
> 
> This feels like the criteria is being based on an implementation detail. For what it's worth, I'd say that `font="base"` _does_ in fact apply styles. The component author chose to put them on the `root` class, and let the other font values override that, rather than having a separate `base` class. That's very reasonable, but it could have been written like this instead:
> 
> ```diff
>   root: {
> -    fontFamily: tokens.fontFamilyBase,
>     // ...
>   },
> +  base: { fontFamily: tokens.fontFamilyBase },
>   monospace: { fontFamily: tokens.fontFamilyMonospace },
>   numeric: { fontFamily: tokens.fontFamilyNumeric },
> ```
> 
> ```diff
>   state.root.className = mergeClasses(
>     styles.root,
> +    state.font === 'base' && styles.base,
>     state.font === 'monospace' && styles.monospace,
>     state.font === 'numeric' && styles.numeric,
>     // ...
>   );
> ```
> 
> That's not to say that's a preferable way to do it, but just to illustrate that it is an implementation detail whether a given value applies styles or not. And implementation details shouldn't drive the API design.

I feel like the point is being missed @behowell. For that specific case you mentioned "base" is **not** going to apply **any** styles - no root styles or specific styles. This is what we're trying to solve with this RFC, default values that do nothing in any way to the component should not have a default value string value as that is misleading to the user and requires documentation reading. This is aggravated by the fact that component props use one string value and others another, so the user gains no reusable information when reading the docs on one of the components.
The point with going with `undefined` is that it's a standard, no documentation is required as it's built into JavaScript development what it "does" (in this case, what it doesn't do).

You're correct when saying that criteria will be based on implementation, but not in the specific way you mention. The rule would be:
- Prop's default value does anything?
  - Yes - Apply a default value
    i.e. appearance, positioning, sizing props
  - No - Keep `undefined`
    i.e. any prop that does _nothing_ by default

This RFC is not attempting to change the behavior of any props whose default value is meaningful, like the default values used in `appearance` properties.


tl;dr: If your prop default does nothing, even considering root defined styles, it gets no default value.

**Comment by layershifter** [comment]
> That is a great summary to it @layershifter, I'll be adding it to the RFC as it's much better explained and to the point. Thanks!

👍 

---

To clarify: I support that change as it has sense, but I think that RFC is too much focused on implementation details in hooks (that are unstable). I would separate these parts and focus on changes to components' props for now as all other changes could be done later.

**Comment by andrefcdias** [comment]
I've gone ahead and removed the hook changes, leaving only the standardization of the value.

File: rfcs/react-components/no-action-prop-value-standardization.md [hunk 10]
```diff
@@ -0,0 +1,168 @@
...
+  numeric: styles.fontNumeric,
+} as const;
+
+state.root.className = mergeClasses(
+  cardClassNames.root,
+  styles.root,
+  fontLookup[font ?? '_default'], // Applying a default to keep usage of a lookup
+  // {...}
+  state.root.className,
+);
```
Comment (layershifter) [inline rfcs/react-components/no-action-prop-value-standardization.md:105]: I would also cut this

**Review by layershifter — APPROVED** [review]
@andrefcdias FYI: I updated the link to preview.

File: rfcs/react-components/no-action-prop-value-standardization.md (hunk 10 above)
Comment (andrefcdias) [reply rfcs/react-components/no-action-prop-value-standardization.md:105]: I feel like the implementation pattern should be discussed to avoid individual flavors across the code. Do you think another RFC should tackle it?

File: rfcs/react-components/no-action-prop-value-standardization.md [hunk 11]
```diff
@@ -0,0 +1,168 @@
...
+    start: styles.alignStart,
+    center: styles.alignCenter,
+    end: styles.alignEnd,
+    justify: styles.alignJustify,
+  } as const;
+
+  const className = mergeClasses(
+    classNames.root,
+    styles.root,
+    alignMap[props.align ?? '_default']
```
[the remark this answers is not in the document]
Comment (andrefcdias) [reply rfcs/react-components/no-action-prop-value-standardization.md:149]: Also a possible implementation pattern and I'm ok with suggesting this instead. Thoughts from the other reviewers?

File: rfcs/react-components/no-action-prop-value-standardization.md (hunk 10 above)
Comment (layershifter) [reply rfcs/react-components/no-action-prop-value-standardization.md:105]: I dropped this comment before the release as was thinking that the RFC could go before 🐱  We could do changes to avoid breaks in future, but it does not matter as v9 was released. Let's leave implementation details and discuss them 👍

Comment (andrefcdias) [reply rfcs/react-components/no-action-prop-value-standardization.md:105]: I'll go ahead and change implementation to reflect that there was no discussion yet so we can get it merged then
